src/frontend/http.rs

```diff
@@ -212,18 +213,24 @@ pub fn cached_read_query_authz(
 /// GET /q/[query hash]
 pub async fn cached_read_query(
     query_hash: String,
-    query: String,
+    raw_query: String,
     if_none_match: Option<String>,
     context: Arc<dyn SeafowlContext>,
 ) -> Result<Response, ApiError> {
     // Ignore dots at the end
     let query_hash = query_hash.split('.').next().unwrap();
 
-    let hash_str = str_to_hex_hash(&query);
+    let decoded_query = percent_decode_str(&raw_query).decode_utf8();
+
+    if decoded_query.is_err() {
+        return Err(ApiError::QueryDecodeError);
+    }
+
+    let hash_str = str_to_hex_hash(&decoded_query.as_ref().unwrap());
```

Comment:
Instead of using `unwrap()`, you should use [the `?` operator](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator) to bail out with an Err value and safely extract the Ok value otherwise:

```suggestion
    let decoded_query = percent_decode_str(&raw_query).decode_utf8().map_err(|_| ApiError::QueryDecodeError)?;

    let hash_str = str_to_hex_hash(decoded_query);
```

(decoded_query should have the type String going ahead, so you don't need `unwrap()` which will panic at runtime if it encounters an Error value)